File: backend/go/backend-go-gin/internal/store/storetest/fake.go
// Package storetest provides an in-memory store.Store so handlers, and the seeder can be
// unit-tested without PostgreSQL.
package storetest

import (
	"context"
	"slices"

	"dev.reference/backend-go-gin/internal/store"
)

// Fake records rows in memory. Any of the *Err fields short-circuits the matching method,
// which is how tests exercise the 500 paths.
type Fake struct {
	Items []store.Item
	Users []store.User

	ListItemsErr  error
	CountItemsErr error
	InsertItemErr error
	FindUserErr   error
	CreateUserErr error

	nextItemID int64
	nextUserID int64
}

// New returns an empty Fake.
func New() *Fake {
	return &Fake{}
}

// WithItem appends a pre-existing item and returns the Fake for chaining.
func (f *Fake) WithItem(name, description string) *Fake {
	f.nextItemID++
	f.Items = append(f.Items, store.Item{ID: f.nextItemID, Name: name, Description: description})
	return f
}

// WithUser appends a pre-existing user and returns the Fake for chaining.
func (f *Fake) WithUser(username, passwordHash string) *Fake {
	f.nextUserID++
	f.Users = append(f.Users, store.User{ID: f.nextUserID, Username: username, PasswordHash: passwordHash})
	return f
}

// ListItems implements store.Store.
func (f *Fake) ListItems(context.Context) ([]store.Item, error) {
	if f.ListItemsErr != nil {
		return nil, f.ListItemsErr
	}
	return slices.Clone(f.Items), nil
}

// CountItems implements store.Store.
func (f *Fake) CountItems(context.Context) (int64, error) {
	if f.CountItemsErr != nil {
		return 0, f.CountItemsErr
	}
	return int64(len(f.Items)), nil
}

// InsertItem implements store.Store.
func (f *Fake) InsertItem(_ context.Context, name, description string) error {
	if f.InsertItemErr != nil {
		return f.InsertItemErr
	}
	f.WithItem(name, description)
	return nil
}

// FindUserByUsername implements store.Store.
func (f *Fake) FindUserByUsername(_ context.Context, username string) (store.User, error) {
	if f.FindUserErr != nil {
		return store.User{}, f.FindUserErr
	}
	for _, user := range f.Users {
		if user.Username == username {
			return user, nil
		}
	}
	return store.User{}, store.ErrUserNotFound
}

// CreateUser implements store.Store.
func (f *Fake) CreateUser(_ context.Context, username, passwordHash string) (store.User, error) {
	if f.CreateUserErr != nil {
		return store.User{}, f.CreateUserErr
	}
	for _, user := range f.Users {
		if user.Username == username {
			return store.User{}, store.ErrDuplicateUsername
		}
	}
	f.WithUser(username, passwordHash)
	return f.Users[len(f.Users)-1], nil
}

File: backend/rust/backend-rust-axum/src/store/fake.rs
use std::sync::Mutex;

use super::{Item, Store, StoreError, User};

/// In-memory store. Any of the `fail_*` flags short-circuits the matching method,
/// which is how tests exercise the 500 / 409 paths.
pub struct FakeStore {
    inner: Mutex<FakeInner>,
}

struct FakeInner {
    items: Vec<Item>,
    users: Vec<User>,
    next_item_id: i64,
    next_user_id: i64,
    list_items_err: bool,
    count_items_err: bool,
    insert_item_err: bool,
    find_user_err: Option<InjectedUserError>,
    create_user_err: Option<InjectedUserError>,
    delete_user_err: bool,
}

#[derive(Clone, Copy)]
enum InjectedUserError {
    Sql,
    Duplicate,
}

impl Default for FakeInner {
    fn default() -> Self {
        Self {
            items: Vec::new(),
            users: Vec::new(),
            next_item_id: 0,
            next_user_id: 0,
            list_items_err: false,
            count_items_err: false,
            insert_item_err: false,
            find_user_err: None,
            create_user_err: None,
            delete_user_err: false,
        }
    }
}

fn db_down() -> sqlx::Error {
    sqlx::Error::Protocol("db down".into())
}

impl FakeStore {
    pub fn new() -> Self {
        Self {
            inner: Mutex::new(FakeInner::default()),
        }
    }

    pub fn with_user(self, username: &str, password_hash: &str) -> Self {
        {
            let mut guard = self.inner.lock().unwrap();
            guard.next_user_id += 1;
            let id = guard.next_user_id;
            guard.users.push(User {
                id,
                username: username.to_string(),
                password_hash: password_hash.to_string(),
            });
        }
        self
    }

    pub fn with_item(self, name: &str, description: &str) -> Self {
        {
            let mut guard = self.inner.lock().unwrap();
            guard.next_item_id += 1;
            let id = guard.next_item_id;
            guard.items.push(Item {
                id,
                name: name.to_string(),
                description: description.to_string(),
            });
        }
        self
    }

    pub fn items(&self) -> Vec<Item> {
        self.inner.lock().unwrap().items.clone()
    }

    pub fn users(&self) -> Vec<User> {
        self.inner.lock().unwrap().users.clone()
    }

    pub fn fail_list_items(&self) {
        self.inner.lock().unwrap().list_items_err = true;
    }

    pub fn fail_count_items(&self) {
        self.inner.lock().unwrap().count_items_err = true;
    }

    pub fn fail_insert_item(&self) {
        self.inner.lock().unwrap().insert_item_err = true;
    }

    pub fn fail_find_user(&self) {
        self.inner.lock().unwrap().find_user_err = Some(InjectedUserError::Sql);
    }

    pub fn fail_create_user(&self) {
        self.inner.lock().unwrap().create_user_err = Some(InjectedUserError::Sql);
    }

    pub fn fail_create_user_duplicate(&self) {
        self.inner.lock().unwrap().create_user_err = Some(InjectedUserError::Duplicate);
    }

    pub fn fail_delete_user(&self) {
        self.inner.lock().unwrap().delete_user_err = true;
    }
}

#[async_trait::async_trait]
impl Store for FakeStore {
    async fn list_items(&self) -> Result<Vec<Item>, sqlx::Error> {
        let guard = self.inner.lock().unwrap();
        if guard.list_items_err {
            return Err(db_down());
        }
        Ok(guard.items.clone())
    }

    async fn count_items(&self) -> Result<i64, sqlx::Error> {
        let guard = self.inner.lock().unwrap();
        if guard.count_items_err {
            return Err(db_down());
        }
        Ok(guard.items.len() as i64)
    }

    async fn insert_item(&self, name: &str, description: &str) -> Result<(), sqlx::Error> {
        let mut guard = self.inner.lock().unwrap();
        if guard.insert_item_err {
            return Err(db_down());
        }
        guard.next_item_id += 1;
        let id = guard.next_item_id;
        guard.items.push(Item {
            id,
            name: name.to_string(),
            description: description.to_string(),
        });
        Ok(())
    }

    async fn find_user_by_username(&self, username: &str) -> Result<User, StoreError> {
        let guard = self.inner.lock().unwrap();
        if let Some(kind) = guard.find_user_err {
            return Err(match kind {
                InjectedUserError::Sql => StoreError::Sql(db_down()),
                InjectedUserError::Duplicate => StoreError::Duplicate,
            });
        }
        guard
            .users
            .iter()
            .find(|user| user.username == username)
            .cloned()
            .ok_or(StoreError::NotFound)
    }

    async fn create_user(&self, username: &str, password_hash: &str) -> Result<User, StoreError> {
        let mut guard = self.inner.lock().unwrap();
        if let Some(kind) = guard.create_user_err {
            return Err(match kind {
                InjectedUserError::Sql => StoreError::Sql(db_down()),
                InjectedUserError::Duplicate => StoreError::Duplicate,
            });
        }
        if guard.users.iter().any(|user| user.username == username) {
            return Err(StoreError::Duplicate);
        }
        guard.next_user_id += 1;
        let user = User {
            id: guard.next_user_id,
            username: username.to_string(),
            password_hash: password_hash.to_string(),
        };
        guard.users.push(user.clone());
        Ok(user)
    }

    async fn delete_user(&self, username: &str) -> Result<(), sqlx::Error> {
        let mut guard = self.inner.lock().unwrap();
        if guard.delete_user_err {
            return Err(db_down());
        }
        guard.users.retain(|user| user.username != username);
        Ok(())
    }
}
